packages/apollo-server/src/utils/ApolloServer.ts

```diff
@@ -0,0 +1,383 @@
+    if (this.subscriptions !== false) {
+      const config =
+        this.subscriptions === true || typeof this.subscriptions === 'undefined'
+          ? {
+              path: this.graphqlEndpoint,
+            }
+          : this.subscriptions;
+      this.createSubscriptionServer(this.http, config);
+    }
+
+    return new Promise((s, f) => {
+      if (this.engine) {
+        this.engine.listen(
+          Object.assign({}, options.engine, {
+            graphqlPaths: [this.graphqlEndpoint],
+            port: options.port,
+            httpServer: this.http,
+          }),
+          () => s(this.engine.engineListeningAddress),
+        );
+        this.engine.on('error', f);
+        return;
+      }
+
+      this.http.listen(options.port, () => {
+        const la: any = this.http.address();
+        // Convert IPs which mean "any address" (IPv4 or IPv6) into localhost
+        // corresponding loopback ip. Note that the url field we're setting is
+        // primarily for consumption by our test suite. If this heuristic is
+        // wrong for your use case, explicitly specify a frontend host (in the
+        // `frontends.host` field in your engine config, or in the `host`
+        // option to ApolloEngine.listen).
+        let hostForUrl = la.address;
+        if (la.address === '' || la.address === '::') hostForUrl = 'localhost';
+        la.url = `http://${joinHostPort(hostForUrl, la.port)}`;
+        s(la);
+      });
+    });
+  }
+
+  public async stop() {
+    if (this.engine) await this.engine.stop();
+    if (this.http) await new Promise(s => this.http.close(s));
+  }
+
+  private createSubscriptionServer(server: HttpServer, config: ListenOptions) {
+    const { onDisconnect, onConnect, keepAlive } = config;
+    SubscriptionServer.create(
+      {
+        schema: this.schema,
+        execute,
+        subscribe,
+        onConnect: onConnect
+          ? onConnect
+          : (connectionParams: Object) => ({ ...connectionParams }),
+        onDisconnect: onDisconnect,
+        onOperation: async (_: string, connection: ExecutionParams) => {
+          connection.formatResponse = (value: ExecutionResult) => ({
+            ...value,
+            errors: value.errors && value.errors.map(err => formatError(err)),
+          });
+          let context: Context = this.context ? this.context : { connection };
+
+          try {
+            context =
+              typeof this.context === 'function'
+                ? await this.context({ connection })
+                : context;
+          } catch (e) {
+            console.error(e);
+            throw e;
+          }
+
+          return { ...connection, context };
+        },
+        keepAlive,
+      },
+      {
+        server,
+        path: this.graphqlEndpoint,
+      },
+    );
+  }
+
+  async request(request: Request) {
+    if (!this) {
+      throw new Error(`It looks like you tried to call this.request but didn't bind it to the parent class. To fix this,
```

Comment:
The "who" is framework implementers when building out their integration. `request` is not a public API for ApolloServer though I couldn't label it as private because implementing classes can't implement a private field from what I could [tell](https://www.typescriptlang.org/play/index.html#src=class%20Foo%20%7B%0D%0A%20%20private%20bar()%7B%7D%0D%0A%7D%0D%0A%0D%0Aclass%20Bar%20extends%20Foo%20%7B%0D%0A%20%20private%20bar()%7B%7D%0D%0A%7D)

Reply:
I feel like the error messages aimed at implementors and at end users could be different then? Like this error message makes it feel like this is an important method worth calling, not an internal detail.  And the error should be more like "you called this method directly on the class but you're only supposed to use it as an option in registerMiddleware", not "you should have bound it"?

Re public vs private, is this what `protected` is for?

Reply:
That's exactly what `protected` is for?